lang/java/reef-bridge-proto-java/src/main/java/org/apache/reef/bridge/client/DriverServiceLauncher.java

```diff
@@ -0,0 +1,203 @@
+      final String cmd = StringUtils.join(launchCommand, ' ');
+      builder.setDriverClientLaunchCommand(cmd);
+      builder.addLocalFiles(driverClientConfigurationFile.getAbsolutePath());
+
+      return launch(driverClientConfigurationProto);
+    } finally {
+      driverClientConfigurationFile.deleteOnExit();
+    }
+  }
+
+  private static IDriverLauncher getLocalDriverServiceLauncher() throws InjectionException {
+    final Configuration localJobSubmissionClientConfig = TANG.newConfigurationBuilder()
+        .bindImplementation(IDriverLauncher.class, LocalLauncher.class)
+        .bindImplementation(IDriverServiceConfigurationProvider.class,
+            GRPCDriverServiceConfigurationProvider.class)
+        .build();
+    return TANG.newInjector(localJobSubmissionClientConfig).getInstance(LocalLauncher.class);
+  }
+
+
+  private static IDriverLauncher getYarnDriverServiceLauncher() throws InjectionException {
+    final Configuration yarnJobSubmissionClientConfig = TANG.newConfigurationBuilder()
+        .bindImplementation(IDriverLauncher.class, YarnLauncher.class)
+        .bindImplementation(IDriverServiceConfigurationProvider.class,
+            GRPCDriverServiceConfigurationProvider.class)
+        .build();
+    return TANG.newInjector(yarnJobSubmissionClientConfig).getInstance(YarnLauncher.class);
+  }
+
+  private static IDriverLauncher getAzureBatchDriverServiceLauncher() throws InjectionException {
+    final Configuration azbatchJobSubmissionClientConfig = TANG.newConfigurationBuilder()
+        .bindImplementation(IDriverLauncher.class, AzureBatchLauncher.class)
+        .bindImplementation(IDriverServiceConfigurationProvider.class,
+            GRPCDriverServiceConfigurationProvider.class)
+        .build();
+    return TANG.newInjector(azbatchJobSubmissionClientConfig).getInstance(AzureBatchLauncher.class);
+  }
+
+  private static LauncherStatus launch(
+      final ClientProtocol.DriverClientConfiguration driverClientConfigurationProto) {
+    try {
+      switch (driverClientConfigurationProto.getRuntimeCase()) {
+      case YARN_RUNTIME:
+        final IDriverLauncher yarnDriverServiceLauncher = getYarnDriverServiceLauncher();
+        return yarnDriverServiceLauncher.launch(driverClientConfigurationProto);
+      case LOCAL_RUNTIME:
+        final IDriverLauncher localDriverServiceLauncher = getLocalDriverServiceLauncher();
+        return localDriverServiceLauncher.launch(driverClientConfigurationProto);
+      case AZBATCH_RUNTIME:
+        final IDriverLauncher azureBatchDriverServiceLauncher = getAzureBatchDriverServiceLauncher();
+        return azureBatchDriverServiceLauncher.launch(driverClientConfigurationProto);
+      default:
+        throw new RuntimeException("Unknown runtime");
+      }
+    } catch (final InjectionException ex) {
+      LOG.log(Level.SEVERE, "Job configuration error", ex);
+      throw new RuntimeException(ex);
+    }
+  }
+
+  /**
+   * Main method that launches the REEF job.
```

Comment:
Does this launch the job or the driver?

Reply:
A job.

Reply:
I see your point now. This is launching the REEF job, which is a DriverService that includes an application DriverClient. I reuse the REEFLauncher to start the DriverService.Driver. Please let me know if you think things could be more clear.